From 33abcd86745a51230a65b5a160d62d7a2cbdea3c Mon Sep 17 00:00:00 2001
From: Alice Mikhaylenko <alicem@gnome.org>
Date: Sat, 15 Mar 2025 00:09:17 +0400
Subject: [PATCH 27/33] Release 1.6.5

---
 NEWS        | 9 +++++++++
 meson.build | 4 ++--
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/NEWS b/NEWS
index 55e8b493..7ef28de6 100644
--- a/NEWS
+++ b/NEWS
@@ -1,3 +1,12 @@
+=============
+Version 1.6.5
+=============
+
+- AdwDialog
+  - Fix :current-breakpoint notifications
+- AdwPreferencesDialog/Window
+  - Fix the search filter expression
+
 =============
 Version 1.6.4
 =============
diff --git a/meson.build b/meson.build
index 9f38a5a9..724312ac 100644
--- a/meson.build
+++ b/meson.build
@@ -1,5 +1,5 @@
 project('libadwaita', 'c',
-          version: '1.6.4',
+          version: '1.6.5',
           license: 'LGPL-2.1-or-later',
     meson_version: '>= 0.59.0',
   default_options: [ 'warning_level=1', 'buildtype=debugoptimized', 'c_std=gnu11' ],
@@ -29,7 +29,7 @@ package_api_name = '@0@-@1@'.format(meson.project_name(), apiversion)
 # to 0. When bumping the second version, set the third one to zero.
 #
 # A lot easier than libtool, right?
-libversion = '0.6.4'
+libversion = '0.6.5'
 # The so major version of the library
 soversion = 0
 
-- 
2.47.1

