From 3a8401cf61c21ca1d49f27f5b6aaddc1c372834d Mon Sep 17 00:00:00 2001
From: Alice Mikhaylenko <alicem@gnome.org>
Date: Sat, 2 Aug 2025 01:42:03 +0400
Subject: [PATCH 09/18] Release 1.6.9

---
 NEWS        | 16 ++++++++++++++++
 meson.build |  4 ++--
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/meson.build b/meson.build
index 8a733039..ab6bf2a7 100644
--- a/meson.build
+++ b/meson.build
@@ -1,5 +1,5 @@
 project('libadwaita', 'c',
-          version: '1.6.8',
+          version: '1.6.9',
           license: 'LGPL-2.1-or-later',
     meson_version: '>= 0.59.0',
   default_options: [ 'warning_level=1', 'buildtype=debugoptimized', 'c_std=gnu11' ],
@@ -29,7 +29,7 @@ package_api_name = '@0@-@1@'.format(meson.project_name(), apiversion)
 # to 0. When bumping the second version, set the third one to zero.
 #
 # A lot easier than libtool, right?
-libversion = '0.6.8'
+libversion = '0.6.9'
 # The so major version of the library
 soversion = 0
 
-- 
2.51.1

